cheatsheets/File_Upload_Cheat_Sheet.md

```diff
@@ -0,0 +1,163 @@
+
+Adding some automation to the review could be helpful, which is a harsh process and well studied before its usage. Some services (_e.g._ Virus Total) provide APIs to scan files against well known malicious file hashes. Some frameworks can check and validate the raw content type and validating it against predefined file types, such as in [ASP.NET Drawing Library](https://docs.microsoft.com/en-us/dotnet/api/system.drawing.imaging.imageformat)
+
+### File Storage Location
+
+The location where the files should be stored must be chosen based on security and business requirements. The following points are set by security priority, and are inclusive:
+
+1. Store the files on a **different host**, which allow for complete segragation of duties between the application serving the user, and the host handling file uploads and their storage.
+2. Store the files **outside the webroot**, where only administrative access is allowed.
+3. Store the files **inside the webroot**, and set them in write permissions only.
+   - If read access is required, setting proper controls is a must (_e.g._ internal IP, authorized user, etc.)
+4. Store the files to a **database**, most preferrably as Blobs in a NoSQL database. The database should be chosen based on the application's requirements.
+
+> Some files are emailed on point, or processed and dismissed, without being stored. It is essential to conduct the security measures discussed in this sheet before doing any actions on them.
+
+### User Permissions
+
+Before any file upload service is accessed, proper validation should occur on two levels for the user uploading a file:
+
+- Authentication level
+  - The user should be a registered user, or an identifiable user, in order to set restrictions and limitations for their upload capabilities
+- Authorization level
+  - Make sure that the user's accesses allow them to upload a file, or multiple files for administrative cases.
+
+### System Permissions
+
+Files should be stored in a way that ensures:
+
+- Allowed system users are the only ones capable of reading the files
+- Required modes only are set for the file
+  - If execution is required, scanning the file before running it is required as a security best practice, to ensure that no macros or hidden scripts are available.
+
+### Upload and Download Limits
+
+The application should set proper size limits for the upload service in order to protect the file storage capacity. The file size limit should be considered after file decompression is conducted, and by using secure methods to calculate zip files size. For more on this, see how to [Safely extract files from ZipInputStream](https://wiki.sei.cmu.edu/confluence/display/java/IDS04-J.+Safely+extract+files+from+ZipInputStream), Java's input stream to handle ZIP files.
+
+It should set proper request limits as well for the download service if available to protect the server from DoS attacks.
+
+## Java Code Snippets
```

Comment:
If we're linking to this, are we confident that it is correct and relevant? Looks a bit out of place when we don't give snippets for any other language.

Reply:
@righettod do think that there are more well known best practices for implementing your practices? I can agree that the code written is not battle tested. It is a way to do it though, and shouldn't hurt, but to follow our best practice, your input is highly appreciated.

Reply:
At the time, I have implemented it to address a client issue with office document / image / pdf and then I have proposed a CS about it to share the code in case in which it can help someone too. I have done that because I did not found a library doing the job.

The code provided do the job for the cases mentioned and is currently limited to this. It's up to you to decide to ignore it if you decide that is not relevant. I propose it as sample fixation in my web assessment for the same cases.